src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala

```diff
@@ -363,48 +363,50 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
-            val givenArgs = p.args.flatten
-            if (givenArgs.length != p.args.length) {
-              error(p, s"partial application is not supported for built-in predicates")
+    case p@PPredConstructor(base, _) => {
+      def wellTypedApp(base: PPredConstructorBase): Messages = miscType(base) match {
+        case FunctionT(args, AssertionT) =>
+          val unappliedPositions = p.args.zipWithIndex.filter(_._1.isEmpty).map(_._2)
+          val givenArgs = p.args.zipWithIndex.filterNot(x => unappliedPositions.contains(x._2)).map(_._1.get)
+          val expectedArgs = args.zipWithIndex.filterNot(x => unappliedPositions.contains(x._2)).map(_._1)
+          if (givenArgs.isEmpty && expectedArgs.isEmpty) {
+            noMessages
+          } else {
+            multiAssignableTo.errors(givenArgs map exprType, expectedArgs)(p) ++
+              p.args.flatMap(x => x.map(isExpr(_).out).getOrElse(noMessages))
+          }
+
+        case abstractT: AbstractType =>
+          // contextual information would be necessary to predict the constructor's return type (i.e. to find type of unapplied arguments)
+          // right now we only support fully applied arguments for built-in predicates
+          val givenArgs = p.args.flatten
+          if (givenArgs.length != p.args.length) {
+            error(p, s"partial application is not supported for built-in predicates")
+          } else {
+            val givenArgTypes = givenArgs map exprType
+            val msgs = abstractT.messages(p, givenArgTypes)
+            if (msgs.nonEmpty) {
+              msgs
             } else {
-              val givenArgTypes = givenArgs map exprType
-              val msgs = abstractT.messages(p, givenArgTypes)
-              if (msgs.nonEmpty) {
-                msgs
-              } else {
-                // the typing function should be defined for these arguments as `msgs` is empty
-                abstractT.typing(givenArgTypes) match {
-                  case FunctionT(args, AssertionT) =>
-                    if (givenArgs.isEmpty && args.isEmpty) {
-                      noMessages
-                    } else {
-                      multiAssignableTo.errors(givenArgs map exprType, args)(p) ++
-                        p.args.flatMap(x => x.map(isExpr(_).out).getOrElse(noMessages))
-                    }
-                  case t => error(p, s"expected function type for resolved AbstractType but got $t")
-                }
+              // the typing function should be defined for these arguments as `msgs` is empty
+              abstractT.typing(givenArgTypes) match {
+                case FunctionT(args, AssertionT) =>
+                  if (givenArgs.isEmpty && args.isEmpty) {
+                    noMessages
+                  } else {
+                    multiAssignableTo.errors(givenArgs map exprType, args)(p) ++
+                      p.args.flatMap(x => x.map(isExpr(_).out).getOrElse(noMessages))
+                  }
```

Comment:
I do not understand this part. AbstractT should be the type of a builtin predicate. As such, I would expect that if it is fully applied that then the final type is AssertionT. Long story short, I expected `case AssertionT => noMessages`

Reply:
I discussed it with linard. As it is written, this should be correct.